### PR TITLE
[AutoDiff upstream] DifferentiationUnittest and some e2e tests

### PR DESCRIPTION
**Pull request by marcrasi**
Adds 2 simple e2e tests and some lit subsitutions and unittest libraries
necessary to support them.

### CODE REVIEW

**Comment by marcrasi** [comment]
Note: depends on https://github.com/apple/swift/pull/30909 for tests to pass, so there's no point running tests before that gets merged.

**Review by dan-zheng — APPROVED** [review]
Exciting that tests pass!

File: test/lit.cfg [hunk 1]
```diff
@@ -1648,6 +1648,22 @@ if not getattr(config, 'target_run_simple_swift', None):
         '%%line-directive %%t/main.swift -- '
         '%s %%t/a.out'
         % (config.target_build_swift, mcp_opt, config.target_codesign, config.target_run))
+    # TODO: Remove when forward mode AD support is robust.
+    config.target_run_simple_swift_forward_mode_differentiation = (
+        '%%empty-directory(%%t) && '
+        '%s %%s -enable-experimental-forward-mode-differentiation -o %%t/a.out %s -module-name main  && '
+        '%s %%t/a.out &&'
+        '%s %%t/a.out'
+        % (config.target_build_swift, mcp_opt, config.target_codesign, config.target_run))
```
Comment (dan-zheng) [inline test/lit.cfg:1657]: I think we can avoid adding ad-hoc `lit` substitutions by using the `%target-run-simple-swift\(([^)]+)\)` parameterized substitution:

https://github.com/apple/swift/blob/770d43028ec6657ecc005d8e0c81c12da821e674/test/lit.cfg#L1713

```swift
// RUN: %target-run-simple-swift(-enable-experimental-forward-mode-differentiation)
```

We could add a similar substitution for `%target-run-simple-swiftgyb(...)` too!

**Review by rxwei — APPROVED** [review]

File: test/AutoDiff/validation-test/simple_math.swift [hunk 2]
```diff
@@ -0,0 +1,408 @@
+// RUN: %target-run-simple-swift
+// NOTE(TF-813): verify that enabling forward-mode does not affect reverse-mode.
+// RUN: %target-run-simple-swift-forward-mode-differentiation
+// RUN: %target-swift-frontend -Xllvm -sil-print-after=differentiation %s -emit-sil -o /dev/null 2>&1 | %FileCheck %s
+// REQUIRES: executable_test
+
+import _Differentiation
+import StdlibUnittest
+import DifferentiationUnittest
+#if os(macOS)
```
Comment (dan-zheng) [inline test/AutoDiff/validation-test/simple_math.swift:10]: Can these ifdef'd `libc` imports be removed? If not, more conditions are necessary for Darwin:
```suggestion
#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
import Darwin.C
#else
import Glibc
#endif
```

If `libc` does need to be imported, perhaps `DifferentiationUnittest` could `@_exported import` it so individual tests don't need to. `DifferentiationUnittest` could `@_exported import _Differentiation`, too.

File: test/lit.cfg (hunk 1 above)
Comment (dan-zheng) [reply test/lit.cfg:1657]: Added `%target-run-simple-swiftgyb(...)` in https://github.com/apple/swift/pull/30918 to enable some other forward-mode differentiation tests.

File: test/AutoDiff/validation-test/simple_math.swift (hunk 2 above)
Comment (marcrasi) [reply test/AutoDiff/validation-test/simple_math.swift]: It works without the libc. I have removed it.

Added `@_exported import _Differentiation`.

File: test/lit.cfg (hunk 1 above)
Comment (marcrasi) [reply test/lit.cfg]: Done.

File: stdlib/private/CMakeLists.txt [hunk 3]
```diff
@@ -5,6 +5,7 @@ endif()
 if(SWIFT_BUILD_SDK_OVERLAY)
   # SwiftPrivateThreadExtras makes use of Darwin/Glibc, which is part of the
   # SDK overlay. It can't be built separately from the SDK overlay.
+  add_subdirectory(DifferentiationUnittest)
```
Comment (marcrasi) [inline stdlib/private/CMakeLists.txt:8]: Oh, does this need `if(SWIFT_ENABLE_EXPERIMENTAL_DIFFERENTIABLE_PROGRAMMING)`? I'll add that here and see what happens.

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
@swift-ci please test linux

**Pull request merged**